File: oving2/ov2.c
#include <pthread.h>
#include <stdio.h>

//gcc -std=gnu99 -Wall -g -o helloworld_c helloworld_c.c -lpthread

int i = 0;

pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;

void *thread1(){
	for(int x = 0; x < 1000000; x++){
		pthread_mutex_lock(&mutex);
		i++;
		pthread_mutex_unlock(&mutex);
	}
	return 0;
}

void *thread2(){
	for(int x = 0; x < 1000000; x++){
		pthread_mutex_lock(&mutex);
		i--;
		pthread_mutex_unlock(&mutex);
	}
	return 0;
}

int main(){
	pthread_t myThread1;
	pthread_t myThread2;

	pthread_create(&myThread1, NULL, thread1, NULL);
	pthread_create(&myThread2, NULL, thread2, NULL);	

	pthread_join(myThread1, NULL);
	pthread_join(myThread2, NULL);
	
	printf("%i\n", i);

	return 0;
}

File: oving2/ov2.py
import threading

i = 0;
lock = threading.Lock()


def thread1():
	global i
	for x in range(1000000):
		with lock:
			i += 1

def thread2():
	global i
	for x in range(1000000):
		with lock:
			i -=1



def main():
	myThread1 = threading.Thread(target = thread1, args = (),)
	myThread2 = threading.Thread(target = thread2, args = (),)
	myThread1.start()
	myThread2.start()
	myThread1.join()
	myThread2.join()
	print(i)
	
main()
